### PR TITLE
Lexical Syntax Organization and Expansion

### PR DESCRIPTION
**Pull request by jessetilro**
Reorganized syntax modules and expanded lexical syntax definitions. Wrote a complete lexical syntax definition for numbers, both implemented and tested according to documentation. Passes on all examples of the official documentation. Numbers still require some work though: need to look into lexical restrictions to enforce longest match, and need to write additional test cases, especially negative ones where the parsing should fail.

### CODE REVIEW

File: ruby.test/syntax/lexical/identifier/IdentifierInvalid.spt [hunk 1]
```diff
@@ -0,0 +1,54 @@
+module IdentifierInvalid
```
Comment (FlXlT) [inline ruby.test/syntax/lexical/identifier/IdentifierInvalid.spt]: What about a convention to call these files `<name>Test` or something?

File: ruby/syntax/lexical/VarName.sdf3 [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
 imports
-	Common
+	Identifier
 	Global
 
 lexical syntax
 
 	VarName = Global
-	VarName = "@"ID // TO DO : no spacing allowed.
+	VarName = "@" ID
```
Comment (FlXlT) [inline ruby/syntax/lexical/VarName.sdf3]: For pretty printing you would not want the space right?

File: ruby/syntax/lexical/Whitespace.sdf3 [hunk 3]
```diff
@@ -0,0 +1,37 @@
+module Whitespace
+
+lexical syntax
+
+  // Layout whitespace
+  LAYOUT = [\ \t\n\r]
+  
+  // Block comment whitespace
```
Comment (FlXlT) [inline ruby/syntax/lexical/Whitespace.sdf3]: Why are _comments_ included in the `Whitespace` file? If we want to get a clean structure if would perhaps be more logical to put this into its own file.

File: ruby.test/syntax/lexical/identifier/IdentifierReservedKeywords.spt [hunk 4]
```diff
@@ -0,0 +1,49 @@
+module IdentifierReservedKeywords
+
+language Ruby
+
+start symbol ID
+
+// Automatically generated from doc/reserved-keywords.txt
```
Comment (FlorisXll) [inline ruby.test/syntax/lexical/identifier/IdentifierReservedKeywords.spt]: Should this not (also) be stated in the actual Identifier definition perhaps? Because it is automatically generated it would nice to document this in some sort of convenient way.

**Review by FlXlT — APPROVED** [review]
Check my comments, looks fine overall.

**Comment by FlXlT** [comment]
Nice work on the numbers. Are we still inconsistent with uppercasing: we have `FNAME` and `VarName`?

File: ruby.test/syntax/lexical/identifier/IdentifierInvalid.spt (hunk 1 above)
Comment (jessetilro) [reply ruby.test/syntax/lexical/identifier/IdentifierInvalid.spt]: Good idea, will fix tomorrow or so

File: ruby/syntax/lexical/VarName.sdf3 (hunk 2 above)
Comment (jessetilro) [reply ruby/syntax/lexical/VarName.sdf3]: **TL;DR** I believe whitespaces between character classes / symbols in lexical syntax rules are completely ignored and are just to distinguish between the different components of the rule.

This is lexical syntax so this is different from what you think. Here it does not make any sense to include layout, since lexical syntax is meant to describe the tokens that the CFG works with and optionally separates by layout (AFAIK). So that space character discussed above does not mean anything. In fact, without the space the rule seemed to break. And above all, with space it is more readable. The spaces in rules like this actually serve that purpose I believe; to separate character classes and symbols et cetera. If you would want to support spaces as part of the token in your lexical syntax rule, you would explicitly add it like this:

```
"@" " " ID
```

Or, maybe to support more general whitespace, something like (don't know if it works):

```
"@" LAYOUT* ID
```

File: ruby/syntax/lexical/Whitespace.sdf3 (hunk 3 above)
Comment (jessetilro) [reply ruby/syntax/lexical/Whitespace.sdf3]: Since comments are also regarded as whitespace. They are not part of the AST. But sure, if you think that helps we can put it in a separate file.

File: ruby/syntax/lexical/VarName.sdf3 (hunk 2 above)
Comment (FlXlT) [reply ruby/syntax/lexical/VarName.sdf3]: Alright, clear.

**Comment by jessetilro** [comment]
Updated test suite modules to have a name ending with "Test". This PR is now ready to be merged.

**Comment by FlXlT** [comment]
Looks good!
